lib/spack/spack/build_environment.py

```diff
@@ -302,6 +302,19 @@ def set_build_environment_variables(pkg, env, dirty=False):
             if '/macports/' in p:
                 env.remove_path('PATH', p)
 
+    # Set environment variables if specified for
+    # the given compiler
+    compiler = pkg.compiler
+    environment = compiler.environment
+    if 'set' in environment:
+        env_to_set = environment['set']
+        for key, value in env_to_set.iteritems():
+            env.set('SPACK_%s' % key, value)
+            env.set('%s' % key, value)
+        # Let shell know which variables to set
+        env_variables = ":".join(env_to_set.keys())
+        env.set('SPACK_ENV_TO_SET', env_variables)
```

Comment:
It may pollute the compiler namespace a little less if you use a single variable for the bundled user environment. e.g. `SPACK_ENV_TO_SET=VAR1=VAULE1:VAR2=VALUE2:...`

I can't actually decide if it matters much so this is minor.  The approach above adds a lot of variables but they're easy to grep for if you run `spack env bash`, so that might be *better* for debugging.  This approach adds fewer names to the compiler's environment.  I'm not sure either is compelling from an environment size perspective -- the OS size limits are typically for the total environment and not per variable.